### ISSUE TITLE
group0 joining in a bootstrapping node is performed with features disabled which may lead to inconsistent schema

### ISSUE THREAD

**Issue by tgrabiec** [scylladb/scylladb#12590]
When replaying schema changes on the bootstrapping node, we can see:

```
DEBUG 2023-01-20 17:54:18,343 [shard 0] schema_tables - adjusting schema_tables mutation due to possible in-progress cluster upgrade
```

This is wrong because it means the new node doesn't have all the features enabled and will drop parts of the mutations which depend on new features. This will result in an inconsistent schema.

One such feature is "CDC".

We can see that the `CDC` feature is enabled but only after schema changes are replayed:

```
INFO  2023-01-20 17:52:21,396 [shard 0] raft_group0 - setup_group0: joining group 0...
...
DEBUG 2023-01-20 17:54:18,343 [shard 0] schema_tables - adjusting schema_tables mutation due to possible in-progress cluster upgrade
...
INFO  2023-01-20 17:54:18,672 [shard 0] raft_group0 - setup_group0: the cluster is ready to use Raft. Finishing.
...
INFO  2023-01-20 17:54:19,056 [shard 0] features - Feature CDC is enabled
```

@kostja

**Comment by kostja** [comment]
@tgrabiec I am missing context. Please explain what you did rather than looked at the logs (size of the cluster, scylla.yaml, or otherwise provide steps to reproduce). Then please explain what you mean "the new code doesn't have all the features enabled" - sounds a bit ambiguous, how it is even possible the new code not having all the features. Then assuming it's true I don't get how you think it will result in an inconsistent schema.

**Comment by tgrabiec** [comment]
I ran this from https://github.com/kbr-scylla/scylladb/commit/53fe1543a727931001416243bb0f78f4f4662d2d:
```
./test.py --mode debug test_many_schema_changes -s
```

It starts a cluster, makes schema changes, and then bootstraps a 4th node:

```python
@pytest.mark.asyncio
async def test_stuff(manager: ManagerClient, random_tables: RandomTables):
    for _ in range(40):
        await random_tables.add_table(ncolumns=5)
    await manager.server_add()
```

We have this thing in scylla called cluster features (see `gms::feature`). They are not always enabled, even though supported. They are enabled when they are supported by all nodes in the cluster, based on gossiped state. They control different things, for example whether certain parts of schema tables can be used. There is a code which drops parts of schema mutations which are not supported yet to protect other nodes from getting confused, and to make sure that the schema digest is consistent during rolling upgrade.

The message comes from there:

```c++
static
mutation
redact_columns_for_missing_features(mutation m, schema_features features) {
    if (features.contains(schema_feature::CDC_OPTIONS) && features.contains(schema_feature::PER_TABLE_PARTITIONERS)) {
        return m;
    }
    if (m.schema()->cf_name() != SCYLLA_TABLES) {
        return m;
    }
    slogger.debug("adjusting schema_tables mutation due to possible in-progress cluster upgrade");
    // The global schema ptr make sure it will be registered in the schema registry.
    global_schema_ptr redacted_schema{scylla_tables(features)};
    m.upgrade(redacted_schema);
    return m;
}
```

In this scenario, all nodes are at the same version, this is not an upgrade scenario. So it's incorrect for the new node to not have all the features enabled.

It's unsafe in general, because the replay works on commands which originated on other nodes, which have the features already enabled. We don't know what will be controlled by the features in the future.

I'll give an example from the past. The MC sstable format added support for a new kind of range tombstones. Which sstable is used, is controlled by a gms::feature in order to support downgrades. If a bootstrapping node received such a mutation while it didn't have the MC feature enabled, it would fail to apply such mutation.

Right now, one thing this can cause that I know of is that the new node will see the wrong schema digest until the feature is enabled.  This is benign though.

**Comment by kostja** [comment]
@tgrabiec but the node which joins an existing cluster with enabled features will fail to join unless it supports all these features. So the node must support all enabled features to join already. Are you saying we're still dropping the schema mutations simply because we are looking at the wrong feature list of the joining node?

**Comment by kbr-scylla** [comment]
> @tgrabiec but the node which joins an existing cluster with enabled features will fail to join unless it supports all these features. So the node must support all enabled features to join already.

The joining node must support the feature, but doesn't necessarily have to have enabled it yet. Enabling a feature happens after observing that every node in the cluster supports it. And apparently it happens in the background and races with joining group 0.

**Comment by kostja** [comment]
For the bootstrapping nodes the check should happen immediately - as the node learns what features are enabled in the cluster. This happens in the gossip shadow round, so immediately after that. 
I am looking at somewhere like here:

```        co_await _gossiper.do_shadow_round(initial_contact_nodes);                                  
        _gossiper.check_knows_remote_features(local_features, loaded_peer_features);                
        _gossiper.check_snitch_name_matches(_snitch.local()->get_name());  ```

**Comment by kostja** [comment]
I think check_knows_remote_features could have a feature listener call to enable these right away, WDYT?

**Comment by tgrabiec** [comment]
The reason why we don't enable the features in the shadow round is that the shadow round may decide that the node should not join the cluster due to various checks. Maybe the cluster name is wrong. If we enabled features at this point, then when you change the config and restart the node it would operate from a state which it shouldn't be in.

**Comment by kostja** [comment]
@tgrabiec I think you're as has been the case in this issue before mixing up upgrade and bootstrap scenarios. When we join an existing cluster we must accept the features enabled by it. There is no negotiation about it.

**Comment by tgrabiec** [comment]
Yes, but the point it that at this point we haven't committed to joining _this_ cluster yet. The safety check I mentioned prevents accidental join of the wrong cluster. Which can happen if you make a mistake and use the wrong config file. Then the admin replaces the yaml with a different seed and restarts the node, attempting to join a different cluster now.

**Comment by kostja** [comment]
... and erases the data directory, too. So what's the issue in persisting the cluster features?

**Comment by kostja** [comment]
moreover, you don't have to persist them in the shadow round. Enabling them is enough. They will be persisted anyway once gossip settles.

**Comment by tgrabiec** [comment]
> ... and erases the data directory, too. 

Currently, we don't tell him to do that.

> So what's the issue in persisting the cluster features?

It's against the current design of the shadow round. What's the issue with persisting them after the shadow round?

> moreover, you don't have to persist them in the shadow round. Enabling them is enough. They will be persisted anyway once gossip settles.

Enabling before persisting is unsafe because it may violate feature monotonicity on the local node in case the node restarts before we persist the feature.

**Comment by kostja** [comment]
> > ... and erases the data directory, too.
> 
> Currently, we don't tell him to do that.

Currently we don't tell the user anything - we don't document the case of "what to do if you mistakenly joined the wrong cluster. BTW, you can just join the right one after that". There is no best practice or an established procedure. 

> > So what's the issue in persisting the cluster features?
> 
> It's against the current design of the shadow round.

This is a recursive argument. We are exactly discussing what to do now. We don't need to persist them during the shadow round either. We can persist them afterwards, we create and persist the host id before we start group0, we can just as well persist the core features before we start group0. 

> > moreover, you don't have to persist them in the shadow round. Enabling them is enough. They will be persisted anyway once gossip settles.
> 
> Enabling before persisting is unsafe because it may violate feature monotonicity on the local node in case the node restarts before we persist the feature.

What you describe is outright impossible:  a node  becomes a member of a cluster with a set of enabled features, then restarts, and has these features turned off in the entire cluster while it was down, so it persists a different set of features, and then it hurts somebody.

It's such a huge stretch of imagination that I have hard time continuing the discussion - it's not productive to discuss hypothetical bugs. This whole issue is hypothetical, too. There is no test case or even a specific scenario how it can hurt us today - it may hurt us sometime in the future, and by this time this code base will change - specifically the features can very well become part of the group0 snapshot, so will be applied first thing when this snapshot is transferred, before the schema. As a consequence it's impossible to assess the impact of the problem to the users. 

If not for a very ill-thought patch for this I would not be spending any more time on this until 5.3.

**Comment by tgrabiec** [comment]
> What you describe is outright impossible: a node becomes a member of a cluster with a set of enabled features, then restarts, and has these features turned off in the entire cluster while it was down, so it persists a different set of features, and then it hurts somebody.

This is not what I described. The features are not turned off in the entire cluster, just on this node while it restarts, until it discovered again that it should enable them. This can cause exactly like the one this issue is about, e.g. the commitlog replay can replay mutations with features disabled, and fail to do so because, for example, memtables write sstables using old, incompatible format. We explicitly added code which ensures monotonicity in a2f5ad432f657253a3a9475be152ee52c05ce30f. So it is a guarantee of the feature service. We should not just take this guarantee back without a good reason.

**Comment by tgrabiec** [comment]
> we create and persist the host id before we start group0, we can just as well persist the core features before we start group0.

Agree about that.

**Comment by kostja** [comment]
OK, as you see from my resistance to this patch, I am happy to fix the problem in pretty much any way that doesn't add an extra dependency to group0 on waiting for gossip to settle. 

I assigned the issue to 5.2 initially since I didn't understand its impact. Now I believe it's pretty safe to move it to 5.3, which I just did. In 5.3, I think, the best way to fix it is to switch the feature storage to a group0 managed table and transfer the features alongside group0 snapshot. 

WDYT?

**Comment by kostja** [comment]
> > What you describe is outright impossible: a node becomes a member of a cluster with a set of enabled features, then restarts, and has these features turned off in the entire cluster while it was down, so it persists a different set of features, and then it hurts somebody.
> 
> This is not what I described. The features are not turned off in the entire cluster, just on this node while it restarts, until it discovered again that it should enable them. This can cause problems exactly like the one this issue is about, e.g. the commitlog replay can replay mutations with features disabled, and fail to do so because, for example, memtables write sstables using old, incompatible format. We explicitly added code which ensures monotonicity in [a2f5ad4](https://github.com/scylladb/scylladb/commit/a2f5ad432f657253a3a9475be152ee52c05ce30f). So it is a guarantee of the feature service. We should not just take this guarantee back without a good reason.

If the invariant that we enable cluster features holds at a subsequent restart, we will enable them again before replay. I agree it's fragile, and I don't think it matters now that the discussion turned elsewhere.

**Comment by tgrabiec** [comment]
> I assigned the issue to 5.2 initially since I didn't understand its impact. Now I believe it's pretty safe to move it to 5.3, which I just did. In 5.3, I think, the best way to fix it is to switch the feature storage to a group0 managed table and transfer the features alongside group0 snapshot.

Managing features in group0 makes sense. Although, it's hard to make definite statements about it without seeing a design doc.

I don't think we should avoid fixing trivial bugs just because we plan to switch to something else later. This creates unnecessary dependencies.

I agree that the priority is low, but the risk of applying the fix is also low, while the risk of delaying the fix is moderate (we will forget about it).

**Comment by tgrabiec** [comment]
I just noticed a fix https://github.com/scylladb/scylladb/pull/12881 which references SCYLLA_KEYSPACES feature. This feature is used during schema merge via extract_scylla_specific_keyspace_info() here:

```c++
    co_await proxy.local().get_db().invoke_on_all([&] (replica::database& db) -> future<> {
        for (auto&& val : created) {
            auto scylla_specific_rs = co_await extract_scylla_specific_keyspace_info(proxy, val);
            auto ksm = create_keyspace_from_schema_partition(val, std::move(scylla_specific_rs));
            co_await db.create_keyspace(ksm, proxy.local().get_erm_factory());
```

```c++

future<lw_shared_ptr<query::result_set>> extract_scylla_specific_keyspace_info(distributed<service::storage_proxy>& proxy, const schema_result_value_type& partition) {
    lw_shared_ptr<query::result_set> scylla_specific_rs;
    if (proxy.local().features().cluster_schema_features().contains<schema_feature::SCYLLA_KEYSPACES>()) {
        ...
    }
    co_return scylla_specific_rs;
```

If the feature is not enabled, the keyspace will be created without scylla-specific options.  This can happen when applying schema changes in group0 replay. If this is the case, then we need to fix this bug right now. 

\cc @xemul @kbr-scylla @kostja

**Comment by kostja** [comment]
@tgrabiec  do you have a test or at least a scenario how you think it can impact our users? I checked the code and I still don't see what exactly is getting broken from the user perspective.

**Comment by tgrabiec** [comment]
The test case is currently only in my imagination. I predict that if you create a keyspace with `storage_type` or `storage_options`,  which you do for S3 storage, and then bootstrap a node, that new node will not see that the storage of the keyspace is set to S3.

It's not a released feature, but our devs will step on it during testing and someone will waste time debugging it sooner or later. And we will need to fix it.

**Comment by kostja** [comment]
This issue is scheduled for 5.3 and Petr will fix it by moving features to Raft. This will happen before S3 storage matures. There is no doubt we need to fix this issue, the question was only whether or not to apply Kamil's fix which changes boot order.

**Comment by tgrabiec** [comment]
Waiting for "features on raft" (which is a nontrivial change still in the speculation phase) will incur higher costs for the organization than applying the fix now. The new discovery demonstrates this. Yesterday you didn't know that there is such a dependency, but the risk was there.

**Comment by kostja** [comment]
@tgrabiec I am easily persuaded when I can assess the impact of this report on the community. A test case or at least some steps to reproduce could help. I do not have the same assessment of cost/benefit as you do.

**Comment by mykaul** [comment]
What's the next step here? Is this a 5.4 material?

**Comment by kostja** [comment]
I believe this is fixed @piodul

**Comment by kbr-scylla** [comment]
IIUC it is fixed in experimental raft-topology mode by @piodul 's joining procedure, but the original issue applies to non-raft-topology-mode.

@kostja should we fix this in non-raft-topology mode as well? From this comment I understand you don't want that:

> This issue is scheduled for 5.3 and Petr will fix it by moving features to Raft.

Features on Raft are only for raft-topology mode.

**Comment by mykaul** [comment]
I *think* it's OK to fix it only for Raft.

**Comment by kbr-scylla** [comment]
Remember that today "Raft" may mean two things:
- 5.2 and 5.4 Raft (i.e. group 0 with schema changes on group 0)
- experimental raft-topology mode (in development) (planned for 6.0 IIUC)

The issue is fixed only for the latter.

**Comment by kostja** [comment]
It's fixed for 6.0, so I am closing this.
